File: 556_next_greater_element/next_greater_element.cpp
/*
 * @Date: 2022-07-04
 * @LastEditors: mengsenwang mengsen_wang@163.com
 * @LastEditTime: 2022-07-04
 * @FilePath: /algorithm/556_next_greater_element/next_greater_element.cpp
 */

#include <cassert>
#include <climits>

class Solution {
 public:
  int nextGreaterElement(int n) {
    int x = n, cnt = 1;
    for (; x >= 10 && x / 10 % 10 >= x % 10; x /= 10) {
      ++cnt;
    }
    x /= 10;
    if (x == 0) {
      return -1;
    }

    int targetDigit = x % 10;
    int x2 = n, cnt2 = 0;
    for (; x2 % 10 <= targetDigit; x2 /= 10) {
      ++cnt2;
    }
    x += x2 % 10 - targetDigit;  // 把 x2 % 10 换到 targetDigit 上

    for (int i = 0; i < cnt; ++i, n /= 10) {  // 反转 n 末尾的 cnt 个数字拼到 x 后
      int d = i != cnt2 ? n % 10 : targetDigit;
      if (x > INT_MAX / 10 || (x == INT_MAX / 10 && d > 7)) {
        return -1;
      }
      x = x * 10 + d;
    }
    return x;
  }
};

int main() {
  assert(Solution().nextGreaterElement(12) == 21);
  assert(Solution().nextGreaterElement(21) == -1);
}

File: rust/src/496_next_greater_element/next_greater_element.rs
/*
 * @Date: 2021-10-26 01:01:42
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-26 01:15:51
 * @FilePath: /algorithm/496_next_greater_element/next_greater_element.rs
 * @Description: file content
 */

struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut m: HashMap<i32, i32> = HashMap::new();
        let mut s: Vec<i32> = vec![0; nums2.len()];
        for i in (0..nums2.len()).rev() {
            while s.len() != 0 && s[s.len() - 1] <= nums2[i] {
                s.pop();
            }
            let t = if s.len() == 0 { -1 } else { s[s.len() - 1] };
            m.insert(nums2[i], t);
            s.push(nums2[i]);
        }
        nums1.iter().map(|x| *m.get(x).unwrap()).collect::<Vec<_>>()
    }
}

fn main() {
    {
        let nums1 = vec![4, 1, 2];
        let nums2 = vec![1, 3, 4, 2];
        let ans = vec![-1, 3, -1];
        assert_eq!(Solution::next_greater_element(nums1, nums2), ans);
    }
    {
        let nums1 = vec![2, 4];
        let nums2 = vec![1, 2, 3, 4];
        let ans = vec![3, -1];
        assert_eq!(Solution::next_greater_element(nums1, nums2), ans);
    }
}
